src/mcpdotnet/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -0,0 +1,249 @@
+﻿using System.Reflection;
+using System.Text.Json;
+using McpDotNet.Configuration;
+using McpDotNet.Protocol.Types;
+using McpDotNet.Server;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace McpDotNet;
+
+/// <summary>
+/// Extension to configure the MCP server with tools
+/// </summary>
+public static partial class McpServerBuilderExtensions
+{
+    /// <summary>
+    /// Adds a tool to the server.
+    /// </summary>
+    /// <typeparam name="TTool">The tool type.</typeparam>
+    /// <param name="builder">The builder instance.</param>
+    public static IMcpServerBuilder WithTool<TTool>(this IMcpServerBuilder builder)
+    {
+        return WithTools(builder, typeof(TTool));
+    }
+    /// <summary>
+    /// Adds all tools marked with <see cref="McpToolTypeAttribute"/> from the current assembly to the server.
+    /// </summary>
+    /// <param name="builder">The builder instance.</param>
+    public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder)
+    {
+        return WithToolsFromAssembly(builder, Assembly.GetCallingAssembly());
+    }
+
+    /// <summary>
+    /// Adds tools to the server.
+    /// </summary>
+    /// <param name="builder">The builder instance.</param>
+    /// <param name="toolTypes">Types with marked methods to add as tools to the server.</param>
+    public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder, params Type[] toolTypes)
+    {
+        ArgumentNullException.ThrowIfNull(toolTypes);
+        if (toolTypes.Length == 0)
+            throw new ArgumentException("At least one tool type must be provided.", nameof(toolTypes));
+
+        var tools = new List<Tool>();
+        Dictionary<string, Func<RequestContext<CallToolRequestParams>, CancellationToken, Task<CallToolResponse>>> callbacks = [];
+
+        foreach (var type in toolTypes)
+        {
+            var methods = type.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
+            foreach (var method in methods)
+            {
+                var attribute = method.GetCustomAttribute<McpToolAttribute>();
+                if (attribute != null)
+                {
+                    var tool = CreateTool(method, attribute);
+                    tools.Add(tool);
+
+                    callbacks.Add(tool.Name, async (request, cancellationToken) => await CallTool(request, method, cancellationToken));
+
+                    // register type because method is not static and so we need an instance
+                    if (!method.IsStatic)
+                        builder.Services.AddScoped(type);
```

Comment:
Why scoped?

Reply:
Because I create a scope when instantiate the type. So then the class can have scoped dependencies like DBContexts etc.